src/coreclr/jit/optimizer.cpp

```diff
@@ -4232,6 +4201,15 @@ void Compiler::optInvertWhileLoop(BasicBlock* block)
         return;
     }
 
+    // It has to be a forward jump. Defer this check until after all the cheap checks
+    // are done, since it iterates forward in the block list looking for bbJumpDest.
+    //  TODO-CQ: Check if we can also optimize the backwards jump as well.
+    //
+    if (fgIsForwardBranch(block) == false)
+    {
+        return;
+    }
+
```

Comment:
What happens if you just remove this check entirely?

If it is useful, why not arrange for blocks to be in bbNum order before this phase runs and just check bbNums?

Reply:
Without this check, it actually works just fine. It turns top-tested, top-entry loops with an unconditional back-edge to top-tested, top-entry loops with a duplicated condition at the bottom and thus a conditional back-edge. So, it reduces the branches in a loop from 2 to 1, at the cost of the duplicated condition.

However, there are diffs (compared to this PR), so it deserves further study. I'll open an issue.

It looks like at this phase location, the block numbers are not guaranteed to be sequential. We could call fgRenumberBlocks, but maybe that's too big a hammer.

Reply:
https://github.com/dotnet/runtime/issues/50204